### DISCUSSION TITLE
Running fief from shell/virtualenv (not in docker)

### DISCUSSION THREAD

**Discussion by yaroslaff** [fief-dev/fief#290] [Q&A] [answered]
I successfully tried fief as docker container, it works, but I want to start it directly, in my virtualenv to explore and understand fief better.

My .env file:
~~~
ENVIRONMENT=development
TELEMETRY_ENABLED=0
SECRET=XXX

# DATABASE_TYPE=POSTGRESQL
DATABASE_TYPE=MYSQL
DATABASE_HOST=localhost
DATABASE_PORT=3306
DATABASE_USERNAME=fief
DATABASE_PASSWORD=fiefpassword
DATABASE_NAME=fief

ROOT_DOMAIN=localhost:8000
FIEF_DOMAIN=localhost:8000

SECRET=XXX....
FIEF_CLIENT_ID=XXX....
FIEF_CLIENT_SECRET=XXX....
ENCRYPTION_KEY=XXX....

FIEF_MAIN_USER_EMAIL=me@example.com
FIEF_MAIN_USER_PASSWORD=_MySecretPassword123
~~~

I do not know, If this is important or not, but I checked my mysql fief database, and I see these tables:
~~~
| admin_api_key        |
| admin_session_tokens |
| alembic_version      |
| workspace_users      |
| workspaces           |
~~~
all of them are empty (0 rows) except alembic_version, which has 'version_num = c8cf076d3ffa'

I run `hatch run dev.server.start` and server starts.
But if I open http://localhost:8000/admin I get redirect to http://localhost:8000/admin/auth/login and getting "Internal Server Error" there.

In server console I see this:
~~~
2023-10-19 01:07:28.004 | INFO     | uvicorn.protocols.http.httptools_impl:send:496 - 127.0.0.1:59956 - "GET /auth/login HTTP/1.1" 500 - {}
Exception in ASGI application
Traceback (most recent call last):
...
...
  File "/home/xenon/tmp/fief/fief/fief/dependencies/fief.py", line 26, in _get_endpoint_url
    url = super()._get_endpoint_url(openid_configuration, field)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/xenon/.local/share/hatch/env/virtual/fief-server/jMP_b3Q_/fief-server/lib/python3.11/site-packages/fief_client/client.py", line 240, in _get_endpoint_url
    return openid_configuration[field]
           ~~~~~~~~~~~~~~~~~~~~^^^^^^^
KeyError: 'authorization_endpoint'
~~~
Should I configure openid_configuration some way? And maybe there is an guide how to start fief in virtualenv? I'm doing it intuitively and probably missing some important steps.

**Comment by fief-bailiff** [comment]
Hail, @yaroslaff 👋

I've noticed you shared secret values: `SECRET`, `SECRET`, `FIEF_CLIENT_ID`, `FIEF_CLIENT_SECRET`, `ENCRYPTION_KEY`. Those are highly sensitive and you should keep them secret.

For your security, I've taken the liberty to replace them with dummy values.

**Comment by frankie567** [comment]
Hmm 🤔 Configuration looks correct! Did you start from a clean database?

If not, you should check that the main workspace on the `workspaces` table has the domain set to `localhost:8000`.

**Reply by yaroslaff** [reply]
Initially, I started with clean database, just `CREATE DATABASE ...` and `GRANT ...`. I do not know for sure when it got populated with these (mostly empty) tables. Anyway, now I dropped every table to start from scratch, and getting "Internal server error" with console message: `sqlalchemy.exc.ProgrammingError: (pymysql.err.ProgrammingError) (1146, "Table 'fief.workspaces' doesn't exist")`

How can I initialize database for fief?

**Reply by frankie567** [reply, marked as answer]
In development, you have to run the following command to create the main tables:

```sh
hatch run python -m fief.cli migrate
```

Then, you probably want to create the main workspace. You can do it like this:

```sh
hatch run python -m fief.cli workspaces create-main
hatch run python -m fief.cli workspaces create-main-user --user-email anne@bretagne.duchy --user-password herminetincture
```

For reference, this is what we do in the initalization script of the GitHub Codespace: https://github.com/fief-dev/fief/blob/main/.devcontainer/post-create.sh

**Comment by yaroslaff** [comment]
migrate worked well for me (re-created table structure as I described above) but `hatch run python -m fief.cli workspaces create-main` produced error:
~~~
│ ╭─────────────────────────────────────────── locals ───────────────────────────────────────────╮ │
│ │              context = <sqlalchemy.dialects.mysql.mysqldb.MySQLExecutionContext_mysqldb      │ │
│ │                        object at 0x7f6f3ad7cd90>                                             │ │
│ │               cursor = <pymysql.cursors.Cursor object at 0x7f6f3ad7d690>                     │ │
│ │              dialect = <sqlalchemy.dialects.mysql.pymysql.MySQLDialect_pymysql object at     │ │
│ │                        0x7f6f3ce9b050>                                                       │ │
│ │ effective_parameters = immutabledict({})                                                     │ │
│ │          evt_handled = False                                                                 │ │
│ │           parameters = [immutabledict({})]                                                   │ │
│ │                 self = <sqlalchemy.engine.base.Connection object at 0x7f6f3ad7d110>          │ │
│ │            statement = <sqlalchemy.dialects.mysql.base.MySQLDDLCompiler object at            │ │
│ │                        0x7f6f3ad7d6d0>                                                       │ │
│ │        str_statement = 'CREATE SCHEMA `d76baf12-c495-4785-a8b6-89b1b7b45338`'                │ │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────╯ │
~~~

Despite I have  `DATABASE_NAME=fief` in .env file (and I even set this variable in shell and exported it), it tries to create new database with "random" guid-based name (different ever time). And fails, because fief user has not grant to create new databases. (I am attaching full stderr log)

[2.log](https://github.com/fief-dev/fief/files/13057695/2.log)

Why this commands tries to create database with such random name? Is it needed? Somehow it does not honors `DATABASE_NAME` from .env file?

**Reply by frankie567** [reply]
That's expected: Fief creates a schema for each workspace. In the main database, you have tables managing the set of workspaces on your instances. Then, for each workspace, there is a dedicated database, each one with a full set of tables. The UUID you see is actually corresponds to the workspace ID.

With PostgreSQL, a schema can be seen as a "sub-database" and is created directly under the specified database.

With MySQL, however, it actually creates a proper database. Hence, you'll either need the `root` user or a user that is allowed to create new databases.

**Comment by yaroslaff** [comment]
This worked, thank you!  
I used mysql docker container and root mysql account.
I've got UI on http://localhost:8000/admin/ and was able to login 
(I had to use mysql docker container)

One thing looks strange, could you please comment it?
~~~~
$ hatch run python -m fief.cli workspaces create-main
Main Fief workspace created
Exception ignored in: <function Connection.__del__ at 0x7fbbb8138fe0>
Traceback (most recent call last):
  File "/home/xenon/.local/share/hatch/env/virtual/fief-server/jMP_b3Q_/fief-server/lib/python3.11/site-packages/aiomysql/connection.py", line 1131, in __del__
    self.close()
  File "/home/xenon/.local/share/hatch/env/virtual/fief-server/jMP_b3Q_/fief-server/lib/python3.11/site-packages/aiomysql/connection.py", line 339, in close
    self._writer.transport.close()
  File "/usr/local/lib/python3.11/asyncio/selector_events.py", line 860, in close
    self._loop.call_soon(self._call_connection_lost, None)
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 761, in call_soon
    self._check_closed()
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 519, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
~~~~

Is this expected? Looks like operation finished successfully, because web UI looks working.

**Reply by frankie567** [reply]
Yes, it's an error I also see in CI with MySQL. I'm not sure where it comes from and if there is something to do on our side, but normally it shouldn't provoke errors.
